### PR TITLE
Fix GDB TCP connection on Windows 

### PR DESCRIPTION
**Pull request by empathicqubit**
On Windows we need to initialize WinSock and use the correct send/recv functions instead of read/write

### CODE REVIEW

**Comment by desertkun** [comment]
Thank you for your interest in the project.

1. Let's replace read/write to recv/send for all systems alike. 
2. Maybe we can do a little `bk.debug("Socket error: %d\n", WSAGetLastError());` for windows too.
3. Please move `WSAStartup` into `void debugger_init()` since gdb can run in mi/2 mode (a.k.a. in IDEs), and should probably work for windows too. And you can ditch the `WSACleanup`.

Also, what emulator are you using?

**Comment by empathicqubit** [comment]
The emulator I'm using is Wabbit. I added GDB support to it, but it's not merged yet. I noticed in Wireshark it wasn't connecting at all, but it was working before when I was using z88dk-gdb from my Linux VM.

**Comment by desertkun** [comment]
Let me know when it has the support, so I can list it on [speccytools.org](https://speccytools.org/). Also you can refer to [gdbserver](https://github.com/speccytools/fuse-for-macosx/blob/7a2c1e7a69150172bcb7c5967002818a518ee051/debugger/gdbserver.c) support for FuseX fork I made, it is compatible with z88dk-gdb. Also you can take a look at [windows variant](https://github.com/speccytools/fuse/blob/gdbserver/debugger/gdbserver.c) but it lags behind. Any issues you find, I would appreciate PRs into windows fork, too.

**Comment by empathicqubit** [comment]
I'm not sure how to fix the msys build. I guess I need to edit a Makefile but idk

**Comment by suborb** [comment]
debugger_gdb_packets.h only includes Winsock when compiled with Visual Studio: https://github.com/z88dk/z88dk/blob/master/src/ticks/debugger_gdb_packets.h#L8

File: src/ticks/debugger.c [hunk 1]
```diff
@@ -13,6 +13,10 @@
 #include <stdarg.h>
 #endif
 
+#ifdef _MSC_VER
```
Comment (empathicqubit) [inline src/ticks/debugger.c:16]: @suborb

Comment (empathicqubit) [reply src/ticks/debugger.c:16]: oh wait. I am dumb

Comment (empathicqubit) [reply src/ticks/debugger.c:16]: @suborb I'm actually a little confused. Do I need winsock for mingw or should I change the debugger_init def to be msc only also?

Comment (suborb) [reply src/ticks/debugger.c:16]: Winsock is needed for all toolchains on windows.

Comment (suborb) [reply src/ticks/debugger.c:16]: I think that means this line: https://github.com/z88dk/z88dk/blob/master/src/ticks/Makefile#L27 needs the now badly named variable `$(GDBLDFLAGS)` adding to the end of it - see the rule for z88dk-gdb

**Comment by empathicqubit** [comment]
Can you help me understand what the i command is doing? It doesn't seem to match the GDB spec. It looks like it is giving the emulator a certain byte offset to run to, but the spec says it should cycle step.

**Comment by desertkun** [comment]
Gdb protocol wise? Yeah I think step over a few bytes

**Comment by empathicqubit** [comment]
@desertkun okay so z88dk does not match the spec? That's fine I just wanted to clarify

**Comment by desertkun** [comment]
It seems you are correct. It should be `i,%d` instead. [Mame](https://github.com/mamedev/mame/blob/master/src/osd/modules/debugger/debuggdbstub.cpp#L1350) does not even support `i`. I think we should just remove that instruction and instead place a temporary breakpoint at specific place since this is what [FuseX gdbserver](https://github.com/speccytools/fuse-for-macosx/blob/master/debugger/gdbserver.c#L397) does. Neither of them do not follow spec, including [spectranet-gdbserver](https://github.com/speccytools/spectranet-gdbserver/blob/master/src/server.c#L134)

What complicates things, it's hard to detect a specific version of [fuse](https://github.com/speccytools/fuse-for-macosx/blob/master/debugger/arch.h#L7) as I wrote dull manifest (and so does [mame](https://github.com/mamedev/mame/blob/master/src/osd/modules/debugger/debuggdbstub.cpp#L695)).

I we can simply bump that version, to `1.1`, and send `i,%` for 1.1 or version missing and `i%` for '1.0'.

But that would require an orchestrating of five things at once. So I'd just enjoy the ride and follow suit 😄

**Comment by empathicqubit** [comment]
@desertkun It's probably best to leave it the way it is, but I think what it actually should be doing is advancing the CPU clock n cycles. So that means that you could get stuck in a state "between" an instruction being executed. It sounds really weird to me though.

**Comment by desertkun** [comment]
the xml manifest (`qXfer:features:read:target.xml:0,3fff`) is the feature of "what is was". Only problem is I did not take advantage from it and copy-pasted it 😄 Had too many things on my plate. I've been spamming the owner of [Es.pectrum](https://www.habisoft.com/espectrum/EN.htm) Emulator for gdbserver support but he has no interest in that and doesn't want to open sources either.

I do feel happy that you are following thru with your emulator. Maybe after you implement it, I can convince the author again ("see, you're being outcompeted 😄" )

**Comment by empathicqubit** [comment]
Can you tell me exactly what I need to do to fix this?

https://github.com/z88dk/z88dk/actions/runs/3878501600/jobs/6614801211

**Comment by desertkun** [comment]
I saw network-related issue, reran for the sake of it

**Comment by empathicqubit** [comment]
> Does z88dk-gdb have any command to set memory other than the file one?

**Comment by suborb** [comment]
> Can you tell me exactly what I need to do to fix this?
> 
> https://github.com/z88dk/z88dk/actions/runs/3878501600/jobs/6614801211

It's "winsock2.h" not "WinSock2.h"

**Comment by empathicqubit** [comment]
Okay. The path I have is C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\um\WinSock2.h

**Comment by suborb** [comment]
> Okay. The path I have is C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\um\WinSock2.h

Gotta love file system case insensitivity..

* mingw: https://github.com/mirror/mingw-w64/blob/master/mingw-w64-headers/include/winsock2.h
* Microsoft docs (example): https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsasetlasterror

**Comment by empathicqubit** [comment]
Yeah. It's funny how it's listed two different ways on the Microsoft page, neither of which is the same as the SDK

**Comment by desertkun** [comment]
Thank you for the effort

**Comment by empathicqubit** [comment]
"No probalo!"

**Pull request merged**